plutus-ledger-api/test/Spec/ScriptDecodeError.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Spec.ScriptDecodeError where
+
+import Codec.CBOR.Extras (DeserialiseFailureInfo (..), DeserialiseFailureReason (..))
+import PlutusCore.Version (plcVersion100)
+import PlutusLedgerApi.Common (ScriptDecodeError (..))
+import PlutusLedgerApi.Common.Versions (PlutusLedgerLanguage (..), conwayPV, vasilPV)
+import Prettyprinter (Pretty (pretty))
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (Assertion, testCase, (@?=))
+
+tests :: TestTree
+tests =
+  testGroup
+    "ScriptDecodeError pretty-printing"
+    [ testCase
+        "CBORDeserialiseError"
+        prettyCBORDeserialiseError
+    , testCase
+        "RemainderError"
+        prettyRemainderError
+    , testCase
+        "LedgerLanguageNotAvailableError"
+        prettyLedgerLanguageNotAvailableError
+    , testCase
+        "PlutusCoreLanguageNotAvailableError"
+        prettyPlutusCoreLanguageNotAvailableError
+    ]
+
+prettyCBORDeserialiseError :: Assertion
+prettyCBORDeserialiseError =
+  show (pretty err)
+    @?= "Failed to deserialise a script: CBOR deserialisation failed \
+        \at the offset 12345 for the following reason: reached the end \
+        \of input while more data was expected."
+ where
+  err =
+    CBORDeserialiseError
+      DeserialiseFailureInfo{dfOffset = 12345, dfReason = EndOfInput}
+
+prettyRemainderError :: Assertion
+prettyRemainderError =
+  show (pretty (RemainderError "bytes"))
+    @?= "Script was successfully deserialised, but 5 more bytes were \
+        \encountered after the script's position."
+
+prettyLedgerLanguageNotAvailableError :: Assertion
+prettyLedgerLanguageNotAvailableError =
+  show (pretty err)
+    @?= "You are attempting to use a PlutusV2 script which is not yet supported \
+        \by the current major protocol version 7. \
+        \The major protocol version that introduces this language is 9. \
+        \To learn more about Plutus Core language versions, see \
+        \https://plutus.readthedocs.io/en/latest/explanations/language-versions.html."
+ where
+  err =
+    LedgerLanguageNotAvailableError
+      { sdeAffectedLang = PlutusV2
+      , sdeIntroPv = conwayPV
+      , sdeThisPv = vasilPV
+      }
+
+prettyPlutusCoreLanguageNotAvailableError :: Assertion
+prettyPlutusCoreLanguageNotAvailableError =
+  show (pretty err)
+    @?= "You are attempting to use a Plutus Core version 1.0.0 script with the major \
```

Comment:
"use a Plutus Core version 1.0.0 script" sounds bizarre. How about: "Your script has a Plutus Core version of 1.0.0. This is not supported in PlutusV2 and major protocol version 7".

And this message can be further improved, to point out whether it is because of PlutusV2 or because of major protocol version 7 (or both). Please add a TODO for this.

Reply:
Thanks for the better wording!